### PR TITLE
Add RetryStrategy for S3 file system

### PR DESCRIPTION
**Pull request by yma11**
This PR add `RetryStrategy` support for s3 file system, thus it will retry when connection fails. It also upgrade aws sdk to `1.11.321` which  supports `AdaptiveRetryStrategy` which user may choose to use.

For `RetryStrategy`, 2 configs are added:

`hive.s3.max-attempts`: Maximum attempts for connections to a single http client.
`hive.s3.retry-mode`: 'standard'， 'adaptive' or `legacy`, client will be created w/o retrystrategy if it's empty.

### CODE REVIEW

**Comment by yma11** [comment]
@majetideepak Can you help review this PR? cc @FelixYBW

File: velox/docs/configs.rst
Comment (majetideepak) [inline velox/docs/configs.rst:546]: Can you add some description here on what `standard` and `adaptive` retry strategies are? Thanks!

**Review by majetideepak — COMMENTED** [review]
@yma11 two minor comments. Thanks for this change!

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (yma11) [reply velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:648]: Do you mean keepint the choice that create the client without any `RetryStrategy`?

Comment (yma11) [reply velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp]: Updated a little. Please take a look.

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (FelixYBW) [inline velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:678]: Can you add the check of invalid value of retry mode?

File: velox/docs/configs.rst
Comment (FelixYBW) [inline velox/docs/configs.rst:539]: By default, aws sdk uses 3, right? Let's mention this here

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (yma11) [reply velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:690]: Updated.

**Review by majetideepak — APPROVED** [review]

**Comment by FelixYBW** [comment]
@yma11 When the call client_->HeadObject or client_->GetObject fail, can we get the the retry number from outcome? If so, let's print the number in error message. So user can know it does be retried and may increase retry number.

**Review by majetideepak — CHANGES_REQUESTED** [review]
@yma11 I will approve again once @FelixYBW comment is addressed.

**Comment by yma11** [comment]
> @yma11 When the call client_->HeadObject or client_->GetObject fail, can we get the the retry number from outcome? If so, let's print the number in error message. So user can know it does be retried and may increase retry number.

Hi @FelixYBW There is no interface to get the retry number from outcome. Should be there some info printed during the retry process telling user what's happening? Not sure about it. However, we can get whether the error is retryable based on the response code and give some hints when necessary. Please take a look.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h
Comment (yma11) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:178]: @FelixYBW Please review this change.

File: velox/docs/configs.rst
Comment (jinchengchenghh) [inline velox/docs/configs.rst:542]: As this document describe, default 4 for legacy mode 
```
A default value of 4 for maximum retry attempts,
```
https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-retries.html

File: velox/connectors/hive/HiveConfig.cpp
Comment (jinchengchenghh) [inline velox/connectors/hive/HiveConfig.cpp:123]: does it is `int32_t` in AWS config?

File: velox/docs/configs.rst
Comment (yma11) [reply velox/docs/configs.rst:542]: image

Get this number from code, maybe more reliable?

Comment (FelixYBW) [reply velox/docs/configs.rst:542]: their cpp_sdk and document have some difference. Let's follow sdk

Comment (jinchengchenghh) [reply velox/docs/configs.rst:542]: Maybe different AWS version has different default value

**Comment by yma11** [comment]
> @yma11 I will approve again once @FelixYBW comment is addressed.

Thanks @majetideepak,  I have updated. Can you take a look again? There is a build failure in CI, which is caused by the old version aws sdk doesn't have `GetRetryCount()` API and seems version upgrade in this PR can't be reflected immediately.

**Comment by majetideepak** [comment]
> There is a build failure in CI, which is caused by the old version aws sdk doesn't have `GetRetryCount()` API

@yma11 Let's update the library first in a separate PR (setup-adapters.sh). That will give us the newer CI images.

**Comment by yma11** [comment]
> > There is a build failure in CI, which is caused by the old version aws sdk doesn't have `GetRetryCount()` API
> 
> @yma11 Let's update the library first in a separate PR (changes inside setup-adapters.sh). That will give us the newer CI images with the updated library.

Thanks for suggestion. I have created [9756](https://github.com/facebookincubator/velox/pull/9756) for version upgrade. Will update this PR once it's merged.

**Comment by yma11** [comment]
@majetideepak can you approve this PR again? Thanks.

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:659]: Maybe print the value?
Should 0 be allowed here?
`VELOX_USER_CHECK(
              (maxAttempts.value() >= 0), "Invalid configuration: specified 'hive.s3.max-attempts' value {} is < 0.", maxAttempts.value());`

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:669]: same as above. Should 0 be allowed and specify the value in the error message.

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:680]: same as above. Should 0 be allowed and specify the value in the error message.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:176]: nit: `Request failed after retrying {} times. Try increasing the value of 'hive.s3.max-attempts'.`

**Review by majetideepak — COMMENTED** [review]
@yma11 some minor comments.
Were you able to test this change on your local setup?

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (jinchengchenghh) [inline velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:657]: VELOX_USER_CHECK -> VELOX_USER_CHECK_GT

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (jinchengchenghh) [inline velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:679]: same

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (yma11) [reply velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp]: As checked in [CilentConfiguration](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/client/ClientConfiguration.cpp#L394), it can be `0`. Changed it to use `VELOX_USER_CHECK_GE`. @jinchengchenghh

**Comment by yma11** [comment]
> @yma11 some minor comments. Were you able to test this change on your local setup?

I don't have local env for this test for now. I will add corresponding configs at Gluten side and then test it in AWS.

**Comment by majetideepak** [comment]
> I don't have local env for this test for now. I will add corresponding configs at Gluten side and then test it in AWS.

@yma11 thanks! Can you please confirm and update here? We cannot add CI tests for some of these changes, but the expectation is that the author tests the change locally.

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (FelixYBW) [inline velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:654]: If user doesn't config retryMode but config maxAttempts, does maxAttempts take effect? It should takes effect since we have default retry mode.

Comment (yma11) [reply velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:654]: No. In velox, `maxAttempts` need work together with `retryMode`. If `retryMode` isn't configured, S3 client will be created w/o `RetryStrategy`. But for Gluten, we've set default value for `retryMode`. I updated the doc in this PR.

**Comment by FelixYBW** [comment]
> 
> @yma11 thanks! Can you please confirm and update here? We cannot add CI tests for some of these changes, but the expectation is that the author tests the changes on their end.

Confirmed it works.

**Review by majetideepak — APPROVED** [review]

**Comment by yma11** [comment]
@majetideepak can you help ping anyone who can merge this PR? Thanks.

**Comment by majetideepak** [comment]
@pedroerp can you please help merge this? Thanks!

**Comment by majetideepak** [comment]
@yma11 I pinged Pedro again. Can you please rebase with main? Sorry for the delay.

**Comment by yma11** [comment]
> @yma11 I pinged Pedro again. Can you please rebase with main? Sorry for the delay.

Done. Really thanks for driving merge.

**Pull request closed without merging**
